pkg/asset/machines/azure/azuremachines.go

```diff
@@ -0,0 +1,264 @@
+// Package azure generates Machine objects for azure.
+package azure
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+	capz "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/azure"
+)
+
+// GenerateMachines returns manifests and runtime objects to provision the control plane (including bootstrap, if applicable) nodes using CAPI.
+func GenerateMachines(platform *azure.Platform, pool *types.MachinePool, userDataSecret string, clusterID string, role string, capabilities map[string]string, useImageGallery bool, userTags map[string]string, hyperVGen string, subnet string, resourceGroup string) ([]*asset.RuntimeFile, error) {
+	if poolPlatform := pool.Platform.Name(); poolPlatform != azure.Name {
+		return nil, fmt.Errorf("non-Azure machine-pool: %q", poolPlatform)
+	}
+	mpool := pool.Platform.Azure
+
+	total := int64(1)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+
+	if len(mpool.Zones) == 0 {
+		// if no azs are given we set to []string{""} for convenience over later operations.
+		// It means no-zoned for the machine API
+		mpool.Zones = []string{""}
+	}
+	tags, err := CapzTagsFromUserTags(clusterID, userTags)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create machineapi.TagSpecifications from UserTags: %w", err)
+	}
+
+	var image capz.Image
+	osImage := mpool.OSImage
+	if osImage.Publisher != "" {
+		image = capz.Image{
+			Marketplace: &capz.AzureMarketplaceImage{
+				ImagePlan: capz.ImagePlan{
+					Publisher: osImage.Publisher,
+					Offer:     osImage.Offer,
+					SKU:       osImage.SKU,
+				},
+				Version: osImage.Version,
+			},
+		}
+	} else if useImageGallery {
+		// image gallery names cannot have dashes
+		galleryName := strings.Replace(clusterID, "-", "_", -1)
+		id := clusterID
+		if hyperVGen == "V2" {
+			id += "-gen2"
+		}
+		imageID := fmt.Sprintf("/resourceGroups/%s/providers/Microsoft.Compute/galleries/gallery_%s/images/%s/versions/latest", resourceGroup, galleryName, id)
+		image.ID = &imageID
+	} else {
+		imageID := fmt.Sprintf("/resourceGroups/%s/providers/Microsoft.Compute/images/%s", resourceGroup, clusterID)
+		if hyperVGen == "V2" && platform.CloudName != azure.StackCloud {
+			imageID += "-gen2"
+		}
+		image.ID = &imageID
+	}
+
+	osDisk := capz.OSDisk{
+		OSType:     "Linux",
+		DiskSizeGB: &mpool.DiskSizeGB,
+		ManagedDisk: &capz.ManagedDiskParameters{
+			StorageAccountType: mpool.DiskType,
+		},
+	}
+	ultrassd := mpool.UltraSSDCapability == "Enabled"
+	additionalCapabilities := capz.AdditionalCapabilities{
+		UltraSSDEnabled: &ultrassd,
+	}
```

Comment:
Generally the API is expecting these to be pointers, which you are solving by using `DeepCopy`. Instead, create the variables as pointers and drop DeepCopy:
```suggestion
	additionalCapabilities := &capz.AdditionalCapabilities{
		UltraSSDEnabled: &ultrassd,
	}
```